### PR TITLE
FIX: Make `DisableAndResetAllDevices` handle events while player runs in the background

### PR DESCRIPTION
**Pull request by jfreire-unity**
### Description

https://jira.unity3d.com/browse/ISXB-615

### Changes made

This PR addresses an issue when the player runs in the background (`InputSystem.runInBackground == true` )and `InputSystem.settings.backgroundBehavior = DisableAndResetAllDevices` is set. These settings were previously introduced in #1324.

With this change, `DisableAndResetAllDevices` has a behavior similar to `DisableAndResetNonBackgroundDevices`, where events that are enqueued while the app is in the background will be [discarded based on their type](https://github.com/Unity-Technologies/InputSystem/blob/bcf381095d5108aee26fbaaf2021efd1ccb909ad/Packages/com.unity.inputsystem/InputSystem/InputManager.cs#L3126). [`canEarlyOut`](https://github.com/Unity-Technologies/InputSystem/blob/bcf381095d5108aee26fbaaf2021efd1ccb909ad/Packages/com.unity.inputsystem/InputSystem/InputManager.cs#L2932) no longer depends on `DisableAndResetAllDevices`, which means we won't early out in this situation.

This solves the issue of having to process input events only when the focus was regained which could lead to the wrong triggering of callbacks. Also, the event buffer could grow to big such that we could see errors of type ` Exceeded budget for maximum input event throughput per InputSystem.Update()` is being raised, in case the player stayed out of focus for the time it took to fill the buffer to its max allowed size.

A test was added to test this behavior in the player.

### Notes

_Please write down any additional notes, remove the section if not applicable._

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 1]
```diff
@@ -2934,22 +2932,20 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
...
                 // Early out if there's no events to process.
                 eventBuffer.eventCount == 0
-                || canFlushBuffer ||
+                || canFlushBuffer
                 // If we're in the background and not supposed to process events in this update (but somehow
                 // still ended up here), we're done.
-                ((!gameHasFocus || gameShouldGetInputRegardlessOfFocus) &&
-                    ((m_Settings.backgroundBehavior == InputSettings.BackgroundBehavior.ResetAndDisableAllDevices && updateType != InputUpdateType.Editor)
 #if UNITY_EDITOR
+                || ((!gameHasFocus || gameShouldGetInputRegardlessOfFocus) &&
```
Comment (jamesmcgill) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs:2940]: The 2 lines of comments above should probably move together with this.

**Review by jamesmcgill — APPROVED** [review]
Seems fine, the logic in this part always takes a while to get my head around.

We agreed about letting it flow through in this case so that `ResetAndDisableNonBackgroundDevices` and `ResetAndDisableAllDevices` follow the same path. I think that is still fine but I see that means removing the early out for the player entirely when out of focus and doing the whole update.

I wonder if we could just flush the buffer for the case you had?
I see for Editor it will do that for `ResetAndDisableAllDevices` but not in the Player.

Would something like this work:
```
            var canFlushBuffer =
                    // If out of focus and runInBackground is off (or all devices are disabled).
                    !gameHasFocus
                    && (!m_Runtime.runInBackground || m_Settings.backgroundBehavior == InputSettings.BackgroundBehavior.ResetAndDisableAllDevices)
#if UNITY_EDITOR
                    // .. and ExactlyAsInPlayer is on, discard input.
                    && m_Settings.editorInputBehaviorInPlayMode == InputSettings.EditorInputBehaviorInPlayMode.AllDeviceInputAlwaysGoesToGameView
#endif
                    ;
```

**Comment by jfreire-unity** [comment]
> Seems fine, the logic in this part always takes a while to get my head around.
> 
> We agreed about letting it flow through in this case so that `ResetAndDisableNonBackgroundDevices` and `ResetAndDisableAllDevices` follow the same path. I think that is still fine but I see that means removing the early out for the player entirely when out of focus and doing the whole update.
> 
> I wonder if we could just flush the buffer for the case you had? I see for Editor it will do that for `ResetAndDisableAllDevices` but not in the Player.
> 
> Would something like this work:
> 
> ```
>             var canFlushBuffer =
>                     // If out of focus and runInBackground is off (or all devices are disabled).
>                     !gameHasFocus
>                     && (!m_Runtime.runInBackground || m_Settings.backgroundBehavior == InputSettings.BackgroundBehavior.ResetAndDisableAllDevices)
> #if UNITY_EDITOR
>                     // .. and ExactlyAsInPlayer is on, discard input.
>                     && m_Settings.editorInputBehaviorInPlayMode == InputSettings.EditorInputBehaviorInPlayMode.AllDeviceInputAlwaysGoesToGameView
> #endif
>                     ;
> ```

But if we flush the buffer, we lose events in case devices are removed/added in the background, right? 
The goal of not early out in the player is so that we can only process device configuration events and "flush" all the other events.

**Comment by jfreire-unity** [comment]
> I see for Editor it will do that for ResetAndDisableAllDevices but not in the Player.

Yeah, I only did this to keep our unit tests green for this in the Editor, for now, to avoid breaking some specific use-case 🤷 
I didn't deep dive into the specifics of input event processing in the Editor & Player Mode, to be honest. But probably should.

**Review by ekcoh — APPROVED** [review]
LGTM, as a side note its a pity the Input module isn't aware of this setting. It would make sense to discard data events on lowest level when/if consumer is not emptying the queue. That would have avoided enqueuing the data in the first place.

**Pull request merged**